### PR TITLE
ARROW-16799: [C++] Create a self-pipe abstraction

### PR DESCRIPTION
**Pull request by pitrou**
Also create a FileDescriptor RAII wrapper to automate the chore of closing file descriptors.

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/testing/gtest_util.cc [hunk 1]
```diff
@@ -574,6 +578,40 @@ std::shared_ptr<Array> TweakValidityBit(const std::shared_ptr<Array>& array,
   return MakeArray(data);
 }
 
+// XXX create a testing/io.{h,cc}?
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/testing/gtest_util.cc:581]: Ah, it does feel weird to have it in gtest_util.h, but it's already rather a grab-bag of test helpers so it's not a big deal.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
While most CI failures are unrelated, the test-debian-10-cpp-i386 failure is directly caused by this PR.

File: cpp/src/arrow/util/io_util_test.cc [hunk 2]
```diff
@@ -159,6 +170,282 @@ TEST(WinErrorFromStatus, Basics) {
...
+  ASSERT_TRUE(a.closed());
+  ASSERT_OK(a.Close());
+  ASSERT_OK(a.Close());
+
+  // Assignment
+  ASSERT_OK_AND_ASSIGN(new_fd, NewFileDescriptor());
+  AssertValidFileDescriptor(new_fd);
+  a = FileDescriptor(new_fd);
+  ASSERT_FALSE(a.closed());
+  ASSERT_GT(a.fd(), 2);
```
Comment (cyb70289) [inline cpp/src/arrow/util/io_util_test.cc:208]: Will it fail if `stdin` or `stderr` is closed? Though I don't think it will happen.

**Review by cyb70289 — APPROVED** [review]

File: cpp/src/arrow/util/io_util.h [hunk 3]
```diff
@@ -124,14 +125,46 @@ Result<bool> DeleteFile(const PlatformFilename& file_path, bool allow_not_found
 ARROW_EXPORT
 Result<bool> FileExists(const PlatformFilename& path);
 
+// TODO expose this more publicly to make it available from io/file.h?
+/// A RAII wrapper for a file descriptor.
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/io_util.h:129]: Hmm.. I find 20k references for "a RAII" with Google and only 12k for "an RAII".

File: cpp/src/arrow/util/io_util_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util_test.cc:208]: At least CI is passing :-) We can relax if we encounter a failure IMHO.

**Pull request merged**